### PR TITLE
Move proficiency without level implementation into singleton class

### PR DESCRIPTION
**Pull request by Cararasu**
Centralize all proficiency without level calculation concerning DCs and the proficiency modifier into one central class. Also add interceptor functions, that can be used by modules to change the bonuses for proficiency(0, 2, 4, 6, 8), the proficiency value(+ level), the untrained proficiency bonus from certain feats(e.g. keen recollection) the DC for normal and simple checks and other dc checks(inline DC checks). Also adjusts the AC from transformation spells, so they scale according to the same rules as normal AC.

My reasoning behind those changes are twofold:
Firstly all the traits that give untrained proficiency are extremely overpowered if turning on proficiency without level. This change gives those feats a bonus in between untrained and expert. Also applies to transformation spells.
Secondly I am trying to implement a proper prof. scaling module. To do that hooks into DC and proficiency calculations are necessary, so that it works for both player characters and items.
There is one unresolved issue and that is the xp calculation. I am not sure how to handle that properly.

### CODE REVIEW

**Comment by stwlam** [comment]
Not keen on having these methods at the top level of the actor class. Even with improved support, variant rules are second-class citizens and should be as invisible as possible when not looking for them.

File: packs/data/classfeatures.db/keen-recollection.json [hunk 1]
```diff
@@ -30,7 +30,7 @@
                 ],
                 "selector": "skill-check",
                 "type": "proficiency",
-                "value": "@actor.level"
+                "value": "@actor.untrained_proficiency"
```
[the remark this answers is not in the document]
Comment (Cararasu) [reply packs/data/classfeatures.db/keen-recollection.json:33]: Oh no, but I really like snake case :D Sure I will change it.

Yeah, sry about that. I had some troubles with ´npm run test´.

**Comment by Cararasu** [comment]
> Not keen on having this information at the top level of the actor class. Even with improved support, variant rules are second-class citizens and should be as invisible as possible when not looking for them.

That makes sense. I would love to hide it as deep as possible, so no one needs to think about it but the way it works right now, I don't really see another choice. At least if you want the full support. Once you are in the getter function of level there is no way of distinguishing between who the caller is. Any implementation that handles pwl for feats and spells will need some point where the decision needs to be made and right now this is the only point I can see.

You could make ´level´ a straight up function with some args instead of a getter, but that's an even worse solution imho. Another solution I can think of is to add a modifier which negates the level, but that also seems extremely ugly.

**Comment by CarlosFdez** [comment]
Ugly internal solutions are better than an ugly API. And this is for a variant rule. Modifier to negate the level from all proficiency slug modifiers is plenty good as a solution.

**Comment by Cararasu** [comment]
> Ugly internal solutions are better than an ugly API. And this is for a variant rule. Modifier to negate the level from all proficiency slug modifiers is plenty good as a solution.

Agreed, but I don't really think that applies here. In general a modifier would solve the issue but for certain feats it unfortunately does not.
It would be necessary to track feats like `Martial Experience` internally and apply the negative modifier if one has such a feat. That would require an internal list of feats that add the untrained proficiency level bonus, which is a pretty horrible solution and in case of user generated content would require interaction with that implementation. That sounds like a pretty ugly API to me.
Another solution that I can see is to add the adjusting modifiers inside of the feat definition itself. That seems even worse.
From all the bad solutions this seems to me like it is the better one.

One thing that I have to say though is that my naming of properties was pretty terrible. `adjustedLevel` is a pretty horrible name for what it represents. I changed it to `proficiencyLevelModifier` to make it clear what is going on here.

**Comment by CarlosFdez** [comment]
I guess I will be more blunt. A solution that involves things like @actor.untrainedProficiencyLevelModifier will not be accepted by anyone. Anything that involves solutions that involve changing REs will not really work. Remember that most of these feats and abilities, in the language itself, refer to things like your actor's level. Adding special documentation to our quickstart guides to say "btw when a thing refers to your level for the sake of bonuses, use this property instead" is a complete non-starter.

PWL is a hack, even in its implementation. It not only alters core rules, but also feats and abilities that refer to those core rules. PWL performing adjustments in post is actually in line with how PWL work. I would highly recommended looking into the AdjustModifier rule element, which allows you to change existing modifiers in almost any calculation. If you see gaps, those are probably locations where a modifier isn't properly incorporating adjust modifier, and filling those gaps would be bug fixes.

**Pull request closed without merging**